### PR TITLE
Fix orders badge inconsistency

### LINKED ISSUE #6080: Orders' notification badge can be inconsistent
The orders' notification badge in the app can be inconsistent, to reproduce:

1. Create some orders with processing status.
2. Open the app and check the notifications badge.
3. Go to wp-admin, and remove some of the created orders.
4. Go back to the app, and check that the number is incorrect now, and it will stay incorrect even after refreshing the list of orders.

The cause of this is that we don't remove orders from the app's database, and since the badge is based upon the number of processing orders, it gets out of sync.

We need to find a way to remove deleted orders from the database while keeping the paging logic working.

### PR DESCRIPTION
**Pull request by hichamboushaba**
Closes: #6080

### Description
Currently, the orders' badge is based on the number of orders in the app's database, this causes a major inconsistency due to two things:
1. It doesn't work well with pagination, as it would count only the fetched orders.
2. When an order is deleted, we delete just its summary from the DB, so it stays counted.

This PR attempts to fix these issues by using the API's endpoint `/wc/v3/reports/orders/totals` as the source of truth.

### Testing instructions
It would be great to have a site with a big number of orders to confirm the fix for the pagination issue, you can use this [plugin](https://wordpress.org/plugins/generate-random-orders-for-woocommerce/) for generating random orders.

And before starting the tests, please clear the app's data.

##### Case 1:
1. Open the app.
2. Confirm that the app shows the correct number of unfulfilled orders (orders with the status processing), and compare the number to what Woo Core shows in wp-admin.

##### Case 2: order status change from the app
1. Open the app.
2. Open an order, and toggle its status from `processing` (either enable it or disable it)
3. Confirm the badge is updated accordingly.

##### Case 3: order list fetch
1. Leave the app open
2. Open wp-admin, then toggle an order's status from `processing`.
3. Go back to the app, then refresh the order's list.
4. Confirm the badge is updated accordingly.

##### Case 4: refresh on push notification
1. Leave the app open.
2. Open the store then create an order.
3. Confirm that the badge is updated When the app receives a push notification.
Note: the order might not trigger a change, since orders with payment COD for example are created with status on_hold, in this case, check logcat and confirm the existence of the message: `New order notification received, re-check unfilled orders count`

### Images/gif

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt [hunk 1]
```diff
@@ -19,7 +19,7 @@ class FakeDispatcher : Dispatcher() {
 
     override fun emitChange(changeEvent: Any) {
         // Copy the listeners to avoid any issues because of concurrent changes
-        val listenersCopy = mutableListOf<Any>(listeners.size)
+        val listenersCopy = MutableList<Any>(listeners.size) { }
```
Comment (hichamboushaba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt:22]: This was my mistake when I shared this snippet with Jorge 🤦, I used the wrong API for initializing the List to hold the copy, so it always had the count `1`.

Comment (hichamboushaba) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt]: This was my mistake when working on this implementation 🤦, I used the wrong API, and the resulting list always had 1 as a size, which causes an exception when the number of listeners is more than 1.

Comment (JorgeMucientes) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt]: Interesting TIL

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt [hunk 2]
```diff
@@ -7,73 +7,52 @@ import com.woocommerce.android.R
...
-                    if (count > 0) {
-                        mainView?.showOrderBadge(count)
-                    } else {
-                        mainView?.hideOrderBadge()
-                    }
+            observeProcessingOrdersCount.invoke().collect { count ->
+                if (count != null && count > 0) {
+                    mainView?.showOrderBadge(count)
+                } else {
+                    mainView?.hideOrderBadge()
```
Comment (JorgeMucientes) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/main/MainPresenter.kt:55]: Awesomee! Thanks for refactoring this into much simpler code and getting rid of event bus usage here.

File: WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt [hunk 3]
```diff
@@ -19,7 +19,7 @@ class FakeDispatcher : Dispatcher() {
...
         Collections.copy(listenersCopy, listeners)
 
         listenersCopy.forEach { listener ->
```
Comment (JorgeMucientes) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt:23]: Np. Unrelated to this PR changes but we could simplify this code and probably remove `listenersCopy` by just adding the following line?

```kotlin
        listeners.toList()
            .forEach { listener ->
}
```

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderListViewModelTest.kt [hunk 4]
```diff
@@ -491,7 +491,7 @@ class OrderListViewModelTest : BaseUnitTest() {
         // Reset as we're no interested in previous invocations in this test
         reset(viewModel.activePagedListWrapper)
         viewModel.onNotificationReceived(
-            NotificationReceivedEvent(NotificationChannelType.NEW_ORDER)
+            NotificationReceivedEvent(0L, NotificationChannelType.NEW_ORDER)
```
Comment (JorgeMucientes) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/orders/OrderListViewModelTest.kt]: Super minor np: 
```suggestion
            NotificationReceivedEvent(siteId = 0L, NotificationChannelType.NEW_ORDER)
```

**Review by JorgeMucientes — APPROVED** [review]

**Comment by JorgeMucientes** [comment]
Excellent job @hichamboushaba, the badge count not only works better now but the refactor to the observable count approach is top notch. Thanks for dealing with this. 
Left couple minor np if you want to address them. Otherwise :shipit:

File: WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt (hunk 3 above)
Comment (hichamboushaba) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/FakeDispatcher.kt:23]: Good idea, I wasn't sure `toList` would make a copy of the list, but it seems it's a [documented](https://kotlinlang.org/docs/constructing-collections.html#copy) behavior, thanks for the suggestion @JorgeMucientes, I'll apply it before merging the PR.

**Pull request merged**